### PR TITLE
filebeat/input/httpjson: drain and close response.Body

### PR DESCRIPTION
**Pull request by belimawr**
## What does this PR do?

httpClient.do now reads the whole response.Body into memory and
replaces it with the in-memory copy. This brings some advantages:
 * We can easily ensure the response.Body is always closed
 * We read the response as quick as possible, releasing the connection
   to be reused and avoiding possible timeouts or other errors caused
   by keeping the connection open and blocked.

This came as a improvement on #29816, however I believe it can be merged on its own.

## Why is it important?

It helps to ensure the correctness of the code and to avoid leaking open sockets.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

~~## Author's Checklist~~

~~## How to test this PR locally~~

## Related issues
- #29816

~~## Use cases~~

~~## Screenshots~~

~~## Logs~~

### CODE REVIEW

File: x-pack/filebeat/input/httpjson/request.go [hunk 1]
```diff
@@ -37,8 +37,17 @@ func (c *httpClient) do(stdCtx context.Context, trCtx *transformContext, req *ht
 	if err != nil {
 		return nil, fmt.Errorf("failed to execute http client.Do: %w", err)
 	}
+	defer resp.Body.Close()
+
+	// Read the whole resp.Body so we can release the conneciton.
+	// This implementaion is inspired by httputil.DumpResponse
+	resp.Body, err = drainBody(resp.Body)
```
Comment (cmacknz) [inline x-pack/filebeat/input/httpjson/request.go:44]: Why do we need the `drainBody` function? I have always used something like `io.Copy(ioutil.Discard, resp.Body)` to do this.

File: x-pack/filebeat/input/httpjson/request.go [hunk 2]
```diff
@@ -37,8 +37,17 @@ func (c *httpClient) do(stdCtx context.Context, trCtx *transformContext, req *ht
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
 	if resp.StatusCode > 399 {
-		body, _ := ioutil.ReadAll(resp.Body)
+		body, _ := io.ReadAll(resp.Body)
 		resp.Body.Close()
```
Comment (cmacknz) [inline x-pack/filebeat/input/httpjson/request.go:51]: With the new `defer resp.Body.Close()` above this will close the body twice.

**Review by cmacknz — CHANGES_REQUESTED** [review]

File: x-pack/filebeat/input/httpjson/request.go (hunk 2 above)
Comment (cmacknz) [reply x-pack/filebeat/input/httpjson/request.go:51]: Probably safe, but it is easy to just rely on the deferred call entirely.

File: x-pack/filebeat/input/httpjson/request.go (hunk 1 above)
Comment (belimawr) [reply x-pack/filebeat/input/httpjson/request.go:44]: You mean using a `io.Copy` to copy to a buffer directly?

The main point here is to copy the response body to a in-memory buffer, and use it in the `http.Response` that is returned. I mostly used the same approach as `httputil.DumpResponse` to copy the body into memory and make it available in the `http.Response`. It looks like a neat implementation.

File: x-pack/filebeat/input/httpjson/request.go (hunk 2 above)
Comment (belimawr) [reply x-pack/filebeat/input/httpjson/request.go:51]: Good catch! I removed it on ed127f1ba4

File: x-pack/filebeat/input/httpjson/request.go (hunk 1 above)
Comment (cmacknz) [reply x-pack/filebeat/input/httpjson/request.go:44]: Thanks I needed to read some more of the code to see what is happening. I looked at this a bit too quickly.

The point is to drain the response body as soon as we can, but still allow the `resp.Body` to be read later as if this wasn't taken care of already. I agree this is a neat way to do it.

**Review by cmacknz — APPROVED** [review]

**Review by kvch — APPROVED** [review]

**Pull request merged**